Add three column layout to Page content view

flashbackzoo opened the pull request:
Updated view button styles for new layout

![3-col-1](https://cloud.githubusercontent.com/assets/878176/8420565/b18c433e-1f17-11e5-8285-1335e818fc86.png)

Update breadcrumbs for new layout

![3-col-2](https://cloud.githubusercontent.com/assets/878176/8420567/b52aad96-1f17-11e5-9f90-ea33b08d21dc.png)

Included content filters and view buttons in the SiteTree panel.

![3-col-3](https://cloud.githubusercontent.com/assets/878176/8420569/b8a4dd98-1f17-11e5-8620-3a2460177532.png)

SiteTree panel collapsed

![3-col-4](https://cloud.githubusercontent.com/assets/878176/8420570/bc0c5f10-1f17-11e5-9e21-f1bcdb811252.png)

tractorcow commented:
I would really like to merge this but I think it'd be wise to leave it open for discussion for users to give feedback.

flashbackzoo commented:
@jonom @clarkepaul

tractorcow commented on tests/behat/features/insert-a-link.feature:
```diff
@@ -52,9 +52,9 @@ So that I can link to a external website or a page on my site
+    And I attach the file "testfile.jpg" to "file[Uploads][]" with HTML5
     And I press the "Insert link" button
-    Then the "Content" HTML field should contain "<a href="[file_link,id=1]" target="_blank">awesome</a>"
+    Then the "Content" HTML field should contain "<a href="[file_link,id=3]" target="_blank">awesome</a>"
```
It would be nice if there was a way of wildcarding that file ID, but for the sake of fixing this PR it's fine.

jonom commented:
This is broken now? Shouldn't be visible unless active or hovered over in this view, that's the whole point of the dots.

![pasted_image_29_06_2015_4_19_pm](https://cloud.githubusercontent.com/assets/1079425/8420759/3355db2a-1e7b-11e5-96f6-8941fac54732.png)

jonom commented:
But otherwise I think it's rad :) :+1:

flashbackzoo commented:
Yeah that's broken - I'll fix it.

SpiritLevel commented:
OK! Some feedback from a noob! Looks great as it is, but if you want some really nit-picky ideas, here goes...feel free to completely disregard if not helpful..

(1) the plus sign on the Add New button is pixelated. Same with check marks on Saved, Published.
(2) the "O" on the Home page icon seems misleading. Perhaps an "M" for modified ?
(3) it may seem "O"bsessive of me, but the "O" at the bottom left of the screen doesn't suggest what it does. Perhaps a more representative icon ?
(4) just wondering about "Page name", "URL segment", and "Navigation label" as section descriptions; and the URL given is the full one not just the segment... 
(5) Edit button: "Edit" is not vertically centered in the button. Also, button is too close to end of the URL.
(6) Place "Your Site Name" at top of main window or perhaps in breadcrumbs: "Your Site Name/Home/New Page" ?
(7) security icon? Perhaps a lock instead of three people ?
(8) bar graph instead of a pie chart icon for reports ?
(9) file icon looks like it is only for images...perhaps an icon that represents a more generic file type ?

What an awesome CMS!

flashbackzoo commented:
Fixed badges as pointed out by @jonom 

![3-col-fix1](https://cloud.githubusercontent.com/assets/878176/8421021/67c34a9e-1f1d-11e5-9988-129fc8a61083.png)

Hover:

![3-col-fix2](https://cloud.githubusercontent.com/assets/878176/8421032/79924a72-1f1d-11e5-802a-e0dcc18870fd.png)

jonom commented:
Thanks, that's better :) just FYI @flashbackzoo, you probably saw this and it shouldn't affect your work here but 3.2 now uses a hollow dot (ring) for new pages and a solid dot for modified, and has a much lighter blue for regular hover and active states for consistency. I assume that will be merged down to 3 at some point. https://github.com/silverstripe/silverstripe-framework/pull/4335 

Feels weird that I'm referencing those changes in the Framework project... shouldn't most of the stuff in the `framework/admin` folder be in the CMS project? I'm sure there's a good reason for it, but it does make it a bit hard to know where to go when you want to change stuff in the CMS interface.

flashbackzoo commented:
Hey @SpiritLevel thanks for your feedback.

This pull request is just dealing with splitting the layout into three columns and the styles related to that change. Your points valid - but are outside the scope of this pull request.

There's another pull request that's dealing with the button styles you mentioned https://github.com/silverstripe/silverstripe-framework/pull/4353

@clarkepaul is currently updating some of the icons and he'd be keen for input. Head over to http://silverstripe.uservoice.com/forums/251266-new-features/suggestions/8107791-use-vector-icons-for-all-icons-in-the-cms and leave your feedback :)

SpiritLevel commented:
Sorry about that @flashbackzoo ... Just got on a roll ;)

flashbackzoo commented:
Cheers @jonom yeah I saw that.

The reason for `framework/admin` is allowing `controllers` etc to have a permissions model i.e. only members access `MyAPIController`. If you're providing a permissions model, you need a way to administer members, so there's ModelAdmin.

It's possible this could be split out into a separate module that's required by CMS and Framework in 4.0, but I'm not sure if anyone's given it too much thought.

It is a pain having to make two pull requests for these type of changes, a re-factor of the CSS is definitely going to happen for 4.0, so we'll see what can be moved across to CMS then.

flashbackzoo commented:
@SpiritLevel no worries - always appreciate another set of eyes :)

flashbackzoo commented:
Well - at least the Behat tests are passing on 3 now! I'll see what's going on with the unit tests...

flashbackzoo commented:
Updated CMSMainTest.php to fix the failing build.

The top level (Pages) breadcrumb now appears in the SiteTree panel - that was causing the issue.

![3-col-testfail](https://cloud.githubusercontent.com/assets/878176/8423419/f8b49336-1f3c-11e5-8ce8-f8c70bd2dd30.png)

chillu commented:
This makes the CMS so much more polished, wow! Some niggles:

 * New breadcrumbs break on long page titles: http://take.ms/CSm0V. Please also test that this still works with multiple rows of tabs. Yes, we should avoid that, but tab rows are a reality we have to deal with at the moment :) And check that long "parent" breadcrumb link titles also work.
 * Tree icon isn't enabled when in page edit mode: http://take.ms/piKTB. I can see why (it's switching to full-page tree mode on click), but still confused me. Any ideas?
 * Keyboard focus is hidden when tabbing through to the new UI elements (e.g. tree and list icons). We'll want to keep the CMS at least as accessible as it is at the moment, right?
 * Tree and list icons don't have alternative titles. Same reason as above.

flashbackzoo commented:
Nice spotting. I'll fix those issues tomorrow.

chillu commented:
* When filtering the tree in page edit mode, clicking the tree or list icons doesn't persist the filter.
 * Hiding the search form after submitting makes it much less obvious than in the old implementation that a search filter is currently applied. We have the "Search results. clear" label on there, but its very easy to miss. I'd suggest that we leave the panel expanded, and close it when hitting "clear filter". Also, if you manually close it with filters still applied by clicking on the loupe icon, that icon should be highlighted somehow (blue glow?).
 * The green search button right above a green "add page" button is a bit confusing. Maybe the green is too strong in this context? I suggest we make it a grey background and remove the button styling on "clear filter" in favour of a more subdued link styling (like "more options" in the edit page screen)

flashbackzoo commented:
Thanks - I'll fix the button mode persist issue.

@clarkepaul - Thoughts on the search filters staying expanded?

I'll wait and see what happens with https://github.com/silverstripe/silverstripe-framework/pull/4353 before trying to solve the button issue. But agree - it could be improved.

tractorcow commented:
Passing behat tests on cms, in my travis?

flashbackzoo commented:
@chillu I've fixed some of those issues. Also updated https://github.com/silverstripe/silverstripe-framework/pull/4355

Long breadcrumbs:

![3-col-breadcrumbs](https://cloud.githubusercontent.com/assets/878176/8444805/692d28a6-1fe9-11e5-9be2-7ed491fe0cc3.png)

Button active and focus states:

![3-col-focus](https://cloud.githubusercontent.com/assets/878176/8444761/e0481640-1fe8-11e5-8bbf-1808e64f04a5.png)

I couldn't reproduce the mode persisting issue. Only tree view mode is available in the content editing view, so clicking the list view button will take you back to the top level. That state (for me) persists on subsequent navigations to page content then back to the top level.

chillu commented on tests/behat/features/manage-files.feature:
```diff
@@ -63,6 +63,7 @@ Feature: Manage files
 
   Scenario: I can filter the files list view using name
     Given I expand the content filters
+    And I wait for 2 seconds
```
I see why that's required (animation). But it makes the test less readable, and makes the test even slower to run. Should be solved by having the previous step ("expand the content filters") waiting for animation to finish. Roughly `$this->waitFor('jQuery(':animated').length == 0')`. Do you want to have a crack at it?

flashbackzoo replied:
Yeah good call - I'll update it.

clarkepaul commented:
@chillu if the search stayed expanded we would need to make that a consistent pattern even when there are 20 searchable fields, meaning you would need to scroll to see the results, the user wouldn't even really see any change to the interface. If the search notification isn't obvious enough then I think we need to deal with that rather than showing the search fields. 

Here's an option to consider instead (done pretty fast and crudely so just use as wireframe).

clarkepaul commented:
![silverstripe ui_search](https://cloud.githubusercontent.com/assets/555033/8445285/1160098a-1fef-11e5-8aa4-579206eee1b0.png)

tractorcow commented:
I've restarted the broken builds; Composer timeouts it seems.

chillu commented:
@clarkepaul "... the user wouldn't even really see any change to the interface". Only if we don't make the search panel individually scrollable, and limit its height. Has the disadvantage of pushing the search button out of view, so only feasible if we can fix the search button to the bottom of that panel - which in turn is difficult to achieve with the current layout engine customisations. Even with the new wireframe, its far less obvious what is searched for (or how to find that out) than in 3.1. "Search results for ..." with nice labels would be ideal. Or a "details" link which expands the search panel?

@flashbackzoo Thanks, fixed looking good! The filter not persisting happens in this view when I press the list button: 
![image](https://cloud.githubusercontent.com/assets/111025/8445561/f712ab38-1ff2-11e5-82a6-2f41da280635.png)

clarkepaul commented:
@chillu yup I agree with your statement around what has been searched for (myself and @flashbackzoo already discussed this previously and thought it would be a nice addition) but as always the end of the sprint is looming. If we added the link "Details" for now would that be a suitable improvement (nice idea). I don't really like the idea of a fixed height search panel still.

clarkepaul commented:
![silverstripe ui_search](https://cloud.githubusercontent.com/assets/555033/8445849/979b319e-1ff6-11e5-8ab5-c80785d87aed.png)

chillu commented:
Sweet, yep both of these approaches work for me. Thumbs up for merging as-is for a first step (once the Behat "wait 2 seconds" thing is fixed). Do you think it is feasible to add to your sprint backlog? How do we ensure it doesn't get forgotten about?

flashbackzoo commented:
@chillu fix. see https://github.com/silverstripe/silverstripe-framework/pull/4355 for the context update.

jonom commented:
I think option 2 with the coloured background for the whole panel is a super nice way of indicating a filtered tree

flashbackzoo commented:
ahh yeah - the Behat build is going to fail until the framework PR has been merged...

flashbackzoo commented:
@chillu We've created a story in our backlog to improve the search filters notification :smile:

tractorcow commented:
The tests just went green before it was squashed; So I merged.
